src/storage-client/src/types/sources/data.rs

```diff
@@ -0,0 +1,222 @@
+                                Some(next_pid) => Int32(next_pid),
+                                None => Null,
+                            },
+                            false,
+                        ),
+                    )
+                    .unwrap();
+            }
+            Interval::Point(PartitionId::None) => {}
+        }
+
+        packer.push(Datum::UInt64(partition.timestamp().offset));
+        SourceData(Ok(row))
+    }
+}
+
+impl TryFrom<SourceData> for Partitioned<PartitionId, MzOffset> {
+    type Error = anyhow::Error;
+    fn try_from(data: SourceData) -> Result<Self, Self::Error> {
+        use PartitionId::Kafka;
+
+        let row = data.0.map_err(|_| anyhow!("invalid binding"))?;
+        let mut datums = row.iter();
+        Ok(match (datums.next(), datums.next()) {
+            (Some(Datum::Range(range)), Some(Datum::UInt64(offset))) => {
+                let range = range
+                    .inner
+                    .ok_or_else(|| anyhow!("remap shards do not currently support empty ranges"))?;
+
+                if range.upper.inclusive {
+                    bail!("upper bounds are never inclusive");
+                }
+
+                let lower = range
+                    .lower
+                    .bound
+                    .map(|lower| Kafka(lower.datum().unwrap_int32()));
+
+                let upper = range
+                    .upper
+                    .bound
+                    .map(|upper| Kafka(upper.datum().unwrap_int32()));
+
+                if upper.is_some() && upper < lower {
+                    bail!("range invariant violated: lower must be <= upper, but got lower {:?}, upper {:?}", lower, upper);
+                }
+
+                if range.lower.inclusive {
+                    let lower = match (lower, upper) {
+                        (Some(lower @ Kafka(l)), Some(Kafka(u))) => {
+                            if u.checked_sub(l) != Some(1) {
+                                bail!("point data always written as [lower, lower + 1)");
+                            }
+                            lower
+                        }
+                        (Some(lower @ Kafka(l)), None) => {
+                            if l < i32::MAX {
+                                bail!("point data always written as [lower, lower + 1)");
+                            }
+                            lower
+                        }
+                        _ => bail!("non-partition data packed into column in row"),
+                    };
+
+                    Partitioned::with_partition(lower, MzOffset::from(offset))
+                } else {
+                    if offset != 0 {
+                        bail!("range type ({:?}) with non-0 offset {}", range, offset)
+                    }
+
+                    Partitioned::with_range(lower, upper, MzOffset::from(offset))
+                }
+            }
+            (Some(Datum::UInt64(offset)), None) => {
```

Comment:
We should panic here too. If we use the two implementations of `RemapMigrator` correctly above we should never reach this point

Reply:
This is the line that takes rows of only `Datum::UInt64` into the correct `Partitioned` format in `PersistHandle::next`'s handling of `ListenEvent::Updates`, i.e. how it stores partitionless updates in the `RemapHandle`'s pending batch.